### ISSUE TITLE
[sdk] InvalidInputError when uploading  dsl.pipeline with default values in arguments

### ISSUE THREAD

**Issue by fstetic** [kubeflow/pipelines#8734]
### Environment

*  KFP version:

Local Canonical Kubeflow using [this guide](https://charmed-kubeflow.io/docs/quickstart)
*  KFP SDK version: 

Bottom of KFP UI left sidenav says build version dev_local and the guide states 1.6
*  All dependencies version:

kfp 2.0.0b10
kfp-pipeline-spec 0.1.17
kfp-server-api 2.0.0a6

### Steps to reproduce

Try to upload a pipeline with default values in pipeline function definition. 
E.g.
```
@dsl.pipeline(name='addition-pipeline')
def my_pipeline(a: int, b: int, c: int = 10):
``` 

### Expected result

Pipeline should be uploaded or, if this is the expected behaviour because there exist the option `pipeline_parameters` in compiling, then the [documentation](https://www.kubeflow.org/docs/components/pipelines/v2/compile-a-pipeline/#compile-a-pipeline) should be updated.

Impacted by this bug? Give it a 👍.

**Comment by gkcalat** [comment]
Hi @fstetic!
Thank you for reporting this. Could you please share more details on this: error message, logs, yaml pipeline spec?

**Comment by connor-mccarthy** [comment]
@fstetic, if you `pip install kfp-pipeline-spec==0.1.16` is the error resolved?

**Comment by fstetic** [comment]
@gkcalat sure!

Here is the full error message:
`{"error_message":"Error creating pipeline: Create pipeline failed: InvalidInputError: unknown template format: pipeline spec is invalid","error_details":"Error creating pipeline: Create pipeline failed: InvalidInputError: unknown template format: pipeline spec is invalid"}`
and here is the YAML spec:
```
# PIPELINE DEFINITION
# Name: addition-pipeline
# Inputs:
#    a: int
#    b: int
#    c: int [Default: 10.0]
components:
  comp-addition-component:
    executorLabel: exec-addition-component
    inputDefinitions:
      parameters:
        num1:
          parameterType: NUMBER_INTEGER
        num2:
          parameterType: NUMBER_INTEGER
    outputDefinitions:
      parameters:
        Output:
          parameterType: NUMBER_INTEGER
  comp-addition-component-2:
    executorLabel: exec-addition-component-2
    inputDefinitions:
      parameters:
        num1:
          parameterType: NUMBER_INTEGER
        num2:
          parameterType: NUMBER_INTEGER
    outputDefinitions:
      parameters:
        Output:
          parameterType: NUMBER_INTEGER
deploymentSpec:
  executors:
    exec-addition-component:
      container:
        args:
        - --executor_input
        - '{{$}}'
        - --function_to_execute
        - addition_component
        command:
        - sh
        - -c
        - "\nif ! [ -x \"$(command -v pip)\" ]; then\n    python3 -m ensurepip ||\
          \ python3 -m ensurepip --user || apt-get install python3-pip\nfi\n\nPIP_DISABLE_PIP_VERSION_CHECK=1\
          \ python3 -m pip install --quiet     --no-warn-script-location 'kfp==2.0.0-beta.10'\
          \ && \"$0\" \"$@\"\n"
        - sh
        - -ec
        - 'program_path=$(mktemp -d)

          printf "%s" "$0" > "$program_path/ephemeral_component.py"

          python3 -m kfp.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"

          '
        - "\nimport kfp\nfrom kfp import dsl\nfrom kfp.dsl import *\nfrom typing import\
          \ *\n\ndef addition_component(num1: int, num2: int) -> int:\n    return\
          \ num1 + num2\n\n"
        image: python:3.7
    exec-addition-component-2:
      container:
        args:
        - --executor_input
        - '{{$}}'
        - --function_to_execute
        - addition_component
        command:
        - sh
        - -c
        - "\nif ! [ -x \"$(command -v pip)\" ]; then\n    python3 -m ensurepip ||\
          \ python3 -m ensurepip --user || apt-get install python3-pip\nfi\n\nPIP_DISABLE_PIP_VERSION_CHECK=1\
          \ python3 -m pip install --quiet     --no-warn-script-location 'kfp==2.0.0-beta.10'\
          \ && \"$0\" \"$@\"\n"
        - sh
        - -ec
        - 'program_path=$(mktemp -d)

          printf "%s" "$0" > "$program_path/ephemeral_component.py"

          python3 -m kfp.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"

          '
        - "\nimport kfp\nfrom kfp import dsl\nfrom kfp.dsl import *\nfrom typing import\
          \ *\n\ndef addition_component(num1: int, num2: int) -> int:\n    return\
          \ num1 + num2\n\n"
        image: python:3.7
pipelineInfo:
  name: addition-pipeline
root:
  dag:
    tasks:
      addition-component:
        cachingOptions:
          enableCache: true
        componentRef:
          name: comp-addition-component
        inputs:
          parameters:
            num1:
              componentInputParameter: a
            num2:
              componentInputParameter: b
        taskInfo:
          name: addition-component
      addition-component-2:
        cachingOptions:
          enableCache: true
        componentRef:
          name: comp-addition-component-2
        dependentTasks:
        - addition-component
        inputs:
          parameters:
            num1:
              taskOutputParameter:
                outputParameterKey: Output
                producerTask: addition-component
            num2:
              componentInputParameter: c
        taskInfo:
          name: addition-component-2
  inputDefinitions:
    parameters:
      a:
        parameterType: NUMBER_INTEGER
      b:
        parameterType: NUMBER_INTEGER
      c:
        defaultValue: 10.0
        isOptional: true
        parameterType: NUMBER_INTEGER
schemaVersion: 2.1.0
sdkVersion: kfp-2.0.0-beta.10
```
You made me try uploading it now without `defaultValue: 10.0` and `isOptional: true` and the problem isn't `defaultValue`, it's `isOptional`.

---

Hi @connor-mccarthy! The error isn't resolved, the YAML spec looks identical. Also, pip throws a dependency error when installing `kfp-pipeline-spec 0.1.16`: `kfp 2.0.0b10 requires kfp-pipeline-spec<0.2.0,>=0.1.17, but you have kfp-pipeline-spec 0.1.16 which is incompatible`

**Comment by mereldawu** [comment]
> @gkcalat sure!
> 
> Here is the full error message: `{"error_message":"Error creating pipeline: Create pipeline failed: InvalidInputError: unknown template format: pipeline spec is invalid","error_details":"Error creating pipeline: Create pipeline failed: InvalidInputError: unknown template format: pipeline spec is invalid"}` and here is the YAML spec:
> 
> ```
> # PIPELINE DEFINITION
> # Name: addition-pipeline
> # Inputs:
> #    a: int
> #    b: int
> #    c: int [Default: 10.0]
> components:
>   comp-addition-component:
>     executorLabel: exec-addition-component
>     inputDefinitions:
>       parameters:
>         num1:
>           parameterType: NUMBER_INTEGER
>         num2:
>           parameterType: NUMBER_INTEGER
>     outputDefinitions:
>       parameters:
>         Output:
>           parameterType: NUMBER_INTEGER
>   comp-addition-component-2:
>     executorLabel: exec-addition-component-2
>     inputDefinitions:
>       parameters:
>         num1:
>           parameterType: NUMBER_INTEGER
>         num2:
>           parameterType: NUMBER_INTEGER
>     outputDefinitions:
>       parameters:
>         Output:
>           parameterType: NUMBER_INTEGER
> deploymentSpec:
>   executors:
>     exec-addition-component:
>       container:
>         args:
>         - --executor_input
>         - '{{$}}'
>         - --function_to_execute
>         - addition_component
>         command:
>         - sh
>         - -c
>         - "\nif ! [ -x \"$(command -v pip)\" ]; then\n    python3 -m ensurepip ||\
>           \ python3 -m ensurepip --user || apt-get install python3-pip\nfi\n\nPIP_DISABLE_PIP_VERSION_CHECK=1\
>           \ python3 -m pip install --quiet     --no-warn-script-location 'kfp==2.0.0-beta.10'\
>           \ && \"$0\" \"$@\"\n"
>         - sh
>         - -ec
>         - 'program_path=$(mktemp -d)
> 
>           printf "%s" "$0" > "$program_path/ephemeral_component.py"
> 
>           python3 -m kfp.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"
> 
>           '
>         - "\nimport kfp\nfrom kfp import dsl\nfrom kfp.dsl import *\nfrom typing import\
>           \ *\n\ndef addition_component(num1: int, num2: int) -> int:\n    return\
>           \ num1 + num2\n\n"
>         image: python:3.7
>     exec-addition-component-2:
>       container:
>         args:
>         - --executor_input
>         - '{{$}}'
>         - --function_to_execute
>         - addition_component
>         command:
>         - sh
>         - -c
>         - "\nif ! [ -x \"$(command -v pip)\" ]; then\n    python3 -m ensurepip ||\
>           \ python3 -m ensurepip --user || apt-get install python3-pip\nfi\n\nPIP_DISABLE_PIP_VERSION_CHECK=1\
>           \ python3 -m pip install --quiet     --no-warn-script-location 'kfp==2.0.0-beta.10'\
>           \ && \"$0\" \"$@\"\n"
>         - sh
>         - -ec
>         - 'program_path=$(mktemp -d)
> 
>           printf "%s" "$0" > "$program_path/ephemeral_component.py"
> 
>           python3 -m kfp.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"
> 
>           '
>         - "\nimport kfp\nfrom kfp import dsl\nfrom kfp.dsl import *\nfrom typing import\
>           \ *\n\ndef addition_component(num1: int, num2: int) -> int:\n    return\
>           \ num1 + num2\n\n"
>         image: python:3.7
> pipelineInfo:
>   name: addition-pipeline
> root:
>   dag:
>     tasks:
>       addition-component:
>         cachingOptions:
>           enableCache: true
>         componentRef:
>           name: comp-addition-component
>         inputs:
>           parameters:
>             num1:
>               componentInputParameter: a
>             num2:
>               componentInputParameter: b
>         taskInfo:
>           name: addition-component
>       addition-component-2:
>         cachingOptions:
>           enableCache: true
>         componentRef:
>           name: comp-addition-component-2
>         dependentTasks:
>         - addition-component
>         inputs:
>           parameters:
>             num1:
>               taskOutputParameter:
>                 outputParameterKey: Output
>                 producerTask: addition-component
>             num2:
>               componentInputParameter: c
>         taskInfo:
>           name: addition-component-2
>   inputDefinitions:
>     parameters:
>       a:
>         parameterType: NUMBER_INTEGER
>       b:
>         parameterType: NUMBER_INTEGER
>       c:
>         defaultValue: 10.0
>         isOptional: true
>         parameterType: NUMBER_INTEGER
> schemaVersion: 2.1.0
> sdkVersion: kfp-2.0.0-beta.10
> ```
> 
> You made me try uploading it now without `defaultValue: 10.0` and `isOptional: true` and the problem isn't `defaultValue`, it's `isOptional`.
> 
> Hi @connor-mccarthy! The error isn't resolved, the YAML spec looks identical. Also, pip throws a dependency error when installing `kfp-pipeline-spec 0.1.16`: `kfp 2.0.0b10 requires kfp-pipeline-spec<0.2.0,>=0.1.17, but you have kfp-pipeline-spec 0.1.16 which is incompatible`

Got exactly the same issue (same version but our Kubeflow is on EKS).

**Comment by connor-mccarthy** [comment]
Closing, as duplicate of https://github.com/kubeflow/pipelines/issues/8682. Please continue discussion there.

This is resolved by downgrading to `kfp==2.0.0b9`.

**Comment by ZeynepRuveyda** [comment]
Hi @connor-mccarthy   I am trying to use my local storage but I am getting error. 
kfp version : 2.0.1  however I tried all version of 2.
Kubeflow version : 1.7 
Kubernetes version: 1.24
kfp-kubernetes version : 1.0.0

Here it is pipeline code:  has been taken : [https://kfp-kubernetes.readthedocs.io/en/kfp-kubernetes-1.2.0/](url)
```
from kfp import dsl,compiler
from kfp import kubernetes

@dsl.component
def make_data():
    with open('/home/zeynep/kubeflow_test/hello_world.txt', 'w') as f:
        f.write('my data')
@dsl.component
def read_data():
    with open('/home/zeynep/kubeflow_test/hello_world.txt') as f:
        print(f.read())
@dsl.pipeline
def my_pipeline():
    pvc1 = kubernetes.CreatePVC(
        # can also use pvc_name instead of pvc_name_suffix to use a pre-existing PVC
        pvc_name_suffix='my-pvc',
        access_modes=['ReadWriteOnce'],
        size='5Gi',
        storage_class_name='local-path',
    )
    task1 = make_data()
    # normally task sequencing is handled by data exchange via component inputs/outputs
    # but since data is exchanged via volume, we need to call .after explicitly to sequence tasks
    task2 = read_data().after(task1)
    kubernetes.mount_pvc(
        task1,
        pvc_name=pvc1.outputs['name'],
        mount_path='/home/zeynep/kubeflow_test',
    )
    kubernetes.mount_pvc(   
        task2,
        pvc_name=pvc1.outputs['name'],
        mount_path='/home/zeynep/kubeflow_test',
    )
    # wait to delete the PVC until after task2 completes
    delete_pvc1 = kubernetes.DeletePVC(
        pvc_name=pvc1.outputs['name']).after(task2)
   
compiler.Compiler().compile(my_pipeline, package_path='ex.yaml')  
```

**Error message :**  **{"error_message":"Error creating pipeline version: Create pipeline version failed: InvalidInputError: unknown template format: pipeline spec is invalid","error_details":"Error creating pipeline version: Create pipeline version failed: InvalidInputError: unknown template format: pipeline spec is invalid"}**
